src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -807,63 +807,206 @@ ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query, RangeTblEntry *inse
+				case T_Const:
+				{
+					exprTypeStr = "a constant value";
+				}
 				break;
-			}
 
-			/*
-			 * Reference tables doesn't have a partition column, thus partition columns
-			 * cannot match at all.
-			 */
-			if (PartitionMethod(subqeryTargetEntry->resorigtbl) == DISTRIBUTE_BY_NONE)
-			{
-				partitionColumnsMatch = false;
+				case T_OpExpr:
+				{
+					exprTypeStr = "an operator";
+				}
 				break;
-			}
 
-			if (!IsPartitionColumnRecursive(subqeryTargetEntry->expr, subquery))
-			{
-				partitionColumnsMatch = false;
+				case T_FuncExpr:
+				{
+					FuncExpr *subqueryFunctionExpr =
+						(FuncExpr *) subqeryTargetEntry->expr;
+
+					switch (subqueryFunctionExpr->funcformat)
+					{
+						case COERCE_EXPLICIT_CALL:
+						{
+							exprTypeStr = "a function call";
+						}
+						break;
+
+						case COERCE_EXPLICIT_CAST:
+						{
+							exprTypeStr = "an explicit cast";
+						}
+						break;
+
+						case COERCE_IMPLICIT_CAST:
+						{
+							exprTypeStr = "an implicit cast";
+						}
+						break;
+
+						default:
+						{
+							exprTypeStr = "a function call";
+						}
+						break;
+					}
+				}
+				break;
+
+				case T_Aggref:
+				{
+					exprTypeStr = "an aggregation";
+				}
+				break;
+
+				case T_CaseExpr:
+				{
+					exprTypeStr = "a case expression";
+				}
+				break;
+
+				case T_CoalesceExpr:
+				{
+					exprTypeStr = "a coalesce expression";
+				}
+				break;
+
+				case T_RowExpr:
+				{
+					exprTypeStr = "a row expression";
+				}
+				break;
+
+				case T_MinMaxExpr:
+				{
+					exprTypeStr = "a min/max expression";
+				}
+
+				break;
+
+				case T_CoerceViaIO:
+				{
+					exprTypeStr = "an explicit coercion";
+				}
+
+				break;
+
+				default:
+				{
+					exprTypeStr = exprTypeDefault;
+				}
 				break;
 			}
 
-			partitionColumnsMatch = true;
-			*selectPartitionColumnTableId = subqeryTargetEntry->resorigtbl;
+			errorDetail = psprintf(errorDetailTemplate, exprTypeStr);
 
-			break;
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot plan distributed INSERT INTO .. SELECT query "
+								   "due to partition column restrictions"),
+							errdetail("%s", errorDetail),
+							errhint("Ensure that target partition column matches with a "
+									"plain partition column of a distributed table in "
+									"the subquery.")));
+		}
+
+		/*
+		 * Insert target expression could only be non-var if the select target
+		 * entry does not have the same type (i.e., target column requires casting).
+		 */
+		if (!IsA(targetEntry->expr, Var))
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot plan distributed INSERT INTO .. SELECT query "
+								   "due to data type mismatch of partition columns"),
+							errdetail(
+								"The data type of the target table's partition column "
+								"should exactly match the data type of the "
+								"corresponding target list entry in the subquery.")));
```

Comment:
_target list entry_ is PostgreSQL-hacker speak. For an end user we probably just want _column_.